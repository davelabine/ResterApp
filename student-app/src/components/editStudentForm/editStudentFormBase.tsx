import * as React from 'react';
import { FormGroup, ControlLabel, FormControl, FormControlProps, Thumbnail } from 'react-bootstrap';
import { StudentData, StudentPhoto } from '../../types';

export const BASE_PHOTO_URL: string = 'https://s3-us-west-2.amazonaws.com/';
export const DEFAULT_PHOTO_URL: string = 'https://static.pexels.com/photos/104827/cat-pet-animal-domestic-104827.jpeg';

export interface EditStudentFormBaseProps {
    student: StudentData;
    onFormTextChange: (label: string, filter: String) => void;
    onFormFileChange: (file: File) => void;
    onStudentTextValidate: (label: String) => 'success' | 'warning' | 'error';
}

export interface EditStudentFormBaseState {
    previewURL: string;
}

export class EditStudentFormBase extends React.Component<EditStudentFormBaseProps, EditStudentFormBaseState> {

    constructor(props: EditStudentFormBaseProps) {
        super(props);

        this.state = { previewURL : '' };

        this.handleTextFieldChange = this.handleTextFieldChange.bind(this);
        this.handleFileChange = this.handleFileChange.bind(this);
        this.getPhotoURL = this.getPhotoURL.bind(this);
    }

    public handleFileChange(selectorFiles: FileList | null): void {
        if ( (selectorFiles != null ) && (selectorFiles.length > 0) ) {
            let file: File = selectorFiles[0];

            if (URL.createObjectURL != null) {
                /* A browser DOM depdendency that may not be present */
                this.setState({previewURL: URL.createObjectURL(file)});
            } else {
                this.setState({previewURL: file.name});
            }

            this.props.onFormFileChange(file);
        }
    }

    public handleTextFieldChange(e: React.FormEvent<FormControlProps>): void {
        const label = e.currentTarget.id as string;
        const value = e.currentTarget.value as string;
        console.log('handleTextFieldChange: ' + label + ' ' + value);
        this.props.onFormTextChange(label, value);
    }

    public render() {
        /* Make sure we set a default value for our form controls 
           Otherwise, React throws a warning that we are switching between uncontrolled and controlled components */
        let student = this.props.student;
        if (!student.lastName) { student.lastName = ''; }
        if (!student.firstName) { student.firstName = ''; }
        if (!student.studentId) { student.studentId = ''; }

        return (
            <form>
                <FormGroup
                    validationState={this.props.onStudentTextValidate('lastName')}
                >
                    <ControlLabel>Last Name:</ControlLabel>
                    <FormControl
                        id="lastName"
                        type="text"
                        placeholder="Enter Last Name"
                        value={student.lastName}
                        onChange={this.handleTextFieldChange}
                    />
                    <FormControl.Feedback />
                </FormGroup>
                <FormGroup
                    validationState={this.props.onStudentTextValidate('firstName')}
                >
                    <ControlLabel>First Name:</ControlLabel>
                    <FormControl
                        id="firstName"
                        type="text"
                        placeholder="Enter First Name"
                        value={student.firstName}
                        onChange={this.handleTextFieldChange}
                    />
                    <FormControl.Feedback />
                </FormGroup>                
                <FormGroup
                    validationState={this.props.onStudentTextValidate('studentId')}
                >
                    <ControlLabel>Student ID:</ControlLabel>
                    <FormControl
                        id="studentId"
                        type="text"
                        placeholder="Enter Student Id"
                        value={student.studentId}
                        onChange={this.handleTextFieldChange}
                    />
                    <FormControl.Feedback />
                </FormGroup>
                <FormGroup>
                    <ControlLabel>Photo:</ControlLabel>
                    <input
                        className="preview-photo"
                        type="file"
                        onChange={(e) => this.handleFileChange(e.target.files)}
                    />
                    <Thumbnail 
                        href="#" 
                        alt="171x180" 
                        src={this.getPhotoURL(student.photo as StudentPhoto)}
                    />
                </FormGroup>
            </form>
        );
    }

    private getPhotoURL(photo: StudentPhoto): string {
        let previewURL: string = this.state.previewURL;
        if (previewURL) {
            return previewURL;
        }
        if (photo) {
            return BASE_PHOTO_URL + photo.bucketName + '/' + photo.key;
        }   
        return DEFAULT_PHOTO_URL;
    }
}