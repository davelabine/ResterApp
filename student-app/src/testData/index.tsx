import { StudentData } from '../types';

export const STUDENT_DATA_EMPTY: StudentData = new StudentData();

export const STUDENT_DATA_FAKE: StudentData =  {
    'skey': 'FAKE_KEY',
    'id': 'FAKE_ID',
    'lastName': 'FAKE_NAME',
    'photo': {}
};

export const STUDENT_DATA_BILLY: StudentData =  {
                                                    'skey': '8a80810d5dd89a3f015dd89ac8280000',
                                                    'id': '1234',
                                                    'lastName': 'Billy Bob',
                                                    'photo': {
                                                        'bucketName': 'resterapp-dev',
                                                        'key': 'e93ecbf0-ba96-4290-a906-5ea67e4c4a9f'
                                                    }
                                                };

export const STUDENT_DATA_BILLY_UPDATE: StudentData =  {
                                                    'skey': '8a80810d5dd89a3f015dd89ac8280000',
                                                    'id': '007',
                                                    'lastName': 'James Bond',
                                                    'photo': {}
                                                };

export const LIST_STUDENT_DATA_EMPTY: StudentData[] = [];

export const LIST_STUDENT_DATA_TWO: StudentData[] =  [
    {
        'skey': '8a80810d5dd89a3f015dd89ac8280000',
        'id': '1234',
        'lastName': 'Billy Bob',
        'photo': {
            'bucketName': 'resterapp-dev',
            'key': 'e93ecbf0-ba96-4290-a906-5ea67e4c4a9f'
        }
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a252800001',
        'id': '503074',
        'lastName': 'Albert Russell',
        'photo': {
            'bucketName': 'resterapp-dev',
            'key': 'e93ecbf0-ba96-4290-a906-5ea67e4c4a9f'
        }
    },
];

export const LIST_STUDENT_DATA_BIG: StudentData[] =  [
    {
        'skey': '8a80810d5dd89a3f015dd89ac8280000',
        'id': '1234',
        'lastName': 'Billy Bob',
        'photo': {
            'bucketName': 'resterapp-dev',
            'key': 'e93ecbf0-ba96-4290-a906-5ea67e4c4a9f'
        }
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a252800001',
        'id': '503074',
        'lastName': 'Albert Russell',
        'photo': {
            'bucketName': 'resterapp-dev',
            'key': 'e93ecbf0-ba96-4290-a906-5ea67e4c4a9f'
        }
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a259b2000e',
        'id': '942226',
        'lastName': 'Alfredo Cummings',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a2677f0027',
        'id': '824388',
        'lastName': 'Allen Russell',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a271fb003a',
        'id': '266192',
        'lastName': 'Andre McKenna',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a27f0d0052',
        'id': '932083',
        'lastName': 'Angelica Obrien',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a281c80057',
        'id': '682307',
        'lastName': 'Anna Washington',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a27c52004d',
        'id': '734552',
        'lastName': 'Antonio Fowler',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a25fb80019',
        'id': '615643',
        'lastName': 'Ashley Mcdaniel',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a25897000c',
        'id': '753238',
        'lastName': 'Beatrice Russell',
        'photo': {}
    },
];
