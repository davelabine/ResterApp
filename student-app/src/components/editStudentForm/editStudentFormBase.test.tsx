import * as React from 'react';
import { mount } from 'enzyme';
import { EditStudentFormBase } from './editStudentFormBase';
import * as testData from '../../testData/index';
import toJson from 'enzyme-to-json';

describe('EditStudentformBase', () => {
    it('renders without crashing and matches the last snapshot', () => {
        let mockFormTextChange = jest.fn();
        const list = mount(
                        <EditStudentFormBase
                            student={testData.STUDENT_DATA_EMPTY}
                            onFormTextChange={mockFormTextChange}
                            onFormFileChange={jest.fn()}
                            onStudentTextValidate={jest.fn()}
                        />);
        expect(toJson(list)).toMatchSnapshot();
    });

    it('triggers internal and external form callbacks when a text field is changed', () => {
        const spy = jest.spyOn(EditStudentFormBase.prototype, 'handleTextFieldChange');
        let mockFormTextChange = jest.fn();
        const form = mount(
                        <EditStudentFormBase
                          student={testData.STUDENT_DATA_EMPTY}
                          onFormTextChange={mockFormTextChange}
                          onFormFileChange={jest.fn()}
                          onStudentTextValidate={jest.fn()}
                        />);

        const input = form.find('input'); 
        expect(input.length).toEqual(3);        

        const nameInput = input.find('[id="lastName"]');
        expect(nameInput.length).toEqual(1);
        nameInput.simulate('change', {target: {id: 'lastName', value: 'abc'}});

        const idInput = input.find('[id="id"]');
        expect(idInput.length).toEqual(1);
        idInput.simulate('change', {target: {id: 'id', value: '123'}});

        expect(spy.mock.calls.length).toBe(2);   
        expect(spy.mock.calls[0][0].target.id).toBe('lastName');
        expect(spy.mock.calls[0][0].target.value).toBe('abc');
        expect(spy.mock.calls[1][0].target.id).toBe('id');
        expect(spy.mock.calls[1][0].target.value).toBe('123');

        expect(mockFormTextChange.mock.calls.length).toBe(2);
        expect(mockFormTextChange.mock.calls[0][0]).toBe('lastName');
        expect(mockFormTextChange.mock.calls[1][0]).toBe('id');

        /* Not sure why these dont work...
        expect(mockFormTextChange.mock.calls[0][1]).toBe('abc');
        expect(mockFormTextChange.mock.calls[1][1]).toBe('123');
        */
    });

    it('triggers internal and external form callbacks when the image file is changed', () => {
        
        const spy = jest.spyOn(EditStudentFormBase.prototype, 'handleFileChange');
        let mockFormFileChange = jest.fn();
        const form = mount(
                        <EditStudentFormBase
                          student={testData.STUDENT_DATA_EMPTY}
                          onFormTextChange={jest.fn()}
                          onFormFileChange={mockFormFileChange}
                          onStudentTextValidate={jest.fn()}
                        />);
        const inst = form.instance() as EditStudentFormBase;

        const file = new File(['editStudentFormBase'], 'editStudentFormBase.test.txt', {
            type: 'text/plain',
          });
        form.find('.preview-photo').simulate('change', {target: {files: [file]}});
        expect(spy).toBeCalled();
        expect(inst.state.previewURL.length).toBeGreaterThan(0);
    }); 
});