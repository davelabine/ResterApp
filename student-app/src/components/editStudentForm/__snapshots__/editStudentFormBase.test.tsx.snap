// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditStudentformBase renders without crashing and matches the last snapshot 1`] = `
<EditStudentFormBase
  onFormFileChange={[Function]}
  onFormTextChange={[Function]}
  onStudentTextValidate={[Function]}
  student={
    StudentData {
      "id": "",
      "lastName": "",
    }
  }
>
  <form>
    <FormGroup
      bsClass="form-group"
    >
      <div
        className="form-group has-feedback"
      >
        <ControlLabel
          bsClass="control-label"
          srOnly={false}
        >
          <label
            className="control-label"
          >
            Name:
          </label>
        </ControlLabel>
        <FormControl
          bsClass="form-control"
          componentClass="input"
          id="lastName"
          onChange={[Function]}
          placeholder="Enter Name"
          type="text"
          value=""
        >
          <input
            className="form-control"
            id="lastName"
            onChange={[Function]}
            placeholder="Enter Name"
            type="text"
            value=""
          />
        </FormControl>
        <FormControlFeedback
          bsClass="form-control-feedback"
          bsRole="feedback"
        />
      </div>
    </FormGroup>
    <FormGroup
      bsClass="form-group"
    >
      <div
        className="form-group has-feedback"
      >
        <ControlLabel
          bsClass="control-label"
          srOnly={false}
        >
          <label
            className="control-label"
          >
            ID:
          </label>
        </ControlLabel>
        <FormControl
          bsClass="form-control"
          componentClass="input"
          id="id"
          onChange={[Function]}
          placeholder="Enter Id"
          type="text"
          value=""
        >
          <input
            className="form-control"
            id="id"
            onChange={[Function]}
            placeholder="Enter Id"
            type="text"
            value=""
          />
        </FormControl>
        <FormControlFeedback
          bsClass="form-control-feedback"
          bsRole="feedback"
        />
      </div>
    </FormGroup>
    <FormGroup
      bsClass="form-group"
    >
      <div
        className="form-group"
      >
        <ControlLabel
          bsClass="control-label"
          srOnly={false}
        >
          <label
            className="control-label"
          >
            Photo:
          </label>
        </ControlLabel>
        <input
          className="preview-photo"
          onChange={[Function]}
          type="file"
        />
        <Thumbnail
          alt="171x180"
          bsClass="thumbnail"
          href="#"
          src="https://static.pexels.com/photos/104827/cat-pet-animal-domestic-104827.jpeg"
        >
          <SafeAnchor
            className="thumbnail"
            componentClass="a"
            href="#"
          >
            <a
              className="thumbnail"
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
            >
              <img
                alt="171x180"
                src="https://static.pexels.com/photos/104827/cat-pet-animal-domestic-104827.jpeg"
              />
            </a>
          </SafeAnchor>
        </Thumbnail>
      </div>
    </FormGroup>
  </form>
</EditStudentFormBase>
`;
